modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4HttpRequestStreamIT.java

```diff
@@ -0,0 +1,124 @@
+import org.elasticsearch.rest.BaseRestHandler;
+import org.elasticsearch.rest.RestController;
+import org.elasticsearch.rest.RestHandler;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.RestResponse;
+import org.elasticsearch.rest.RestStatus;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+public class Netty4HttpRequestStreamIT extends ESNetty4IntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return CollectionUtils.concatLists(List.of(RequestContentStreamPlugin.class), super.nodePlugins());
+    }
+
+    @Override
+    protected boolean addMockHttpTransport() {
+        return false; // enable http
+    }
+
+    public void testBasicStream() throws IOException {
+        var totalBytes = 1024 * 1024;
+        var request = new Request("POST", RequestContentStreamPlugin.ROUTE);
+        request.setEntity(new ByteArrayEntity(randomByteArrayOfLength(totalBytes), ContentType.APPLICATION_JSON));
+
+        var respose = getRestClient().performRequest(request);
+        assertEquals(200, respose.getStatusLine().getStatusCode());
+        var gotTotalBytes = new BytesArray(respose.getEntity().getContent().readAllBytes()).utf8ToString();
+        assertEquals("" + totalBytes, gotTotalBytes);
+    }
+
+    public static class RequestContentStreamPlugin extends Plugin implements ActionPlugin {
+
+        static final String ROUTE = "/_test/request-stream/basic";
+        private static final Logger LOGGER = LogManager.getLogger("StreamRestHandler");
+
+        @Override
+        public Collection<RestHandler> getRestHandlers(
+            Settings settings,
+            NamedWriteableRegistry namedWriteableRegistry,
+            RestController restController,
+            ClusterSettings clusterSettings,
+            IndexScopedSettings indexScopedSettings,
+            SettingsFilter settingsFilter,
+            IndexNameExpressionResolver indexNameExpressionResolver,
+            Supplier<DiscoveryNodes> nodesInCluster,
+            Predicate<NodeFeature> clusterSupportsFeature
+        ) {
+            return List.of(new BaseRestHandler() {
+                @Override
+                public String getName() {
+                    return ROUTE;
+                }
+
+                @Override
+                public List<Route> routes() {
+                    return List.of(new Route(RestRequest.Method.POST, ROUTE));
+                }
+
+                @Override
+                protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) {
+                    return channel -> {
+
+                        // Example of handling streamed content
+
+                        var contentStream = request.contentStream();
```

Comment:
We need to do something before `prepareRequest` completes to indicate we will or will not accept a request body before we consume it like this. (see previous comment --> this could be based on the interfaces that the returned consumer implements)